### PR TITLE
parallelization work-in-progress: automatic data-parallel MLP

### PR DESCRIPTION
**Pull request by froystig**
This PR introduces:
* an example program that relies on the auto-parallelization transformation, and
* progress on auto-parallelization machinery to make the example program work.

The transformation parallelizes the computation of the MLP's average loss along the input data's "batch" dimension (examples). The transformation recovers what one would expect: an SPMD computation that first evaluates per-example losses in parallel and then averages them with a `lax.psum` (our all-reduce-sum primitive).

### CODE REVIEW

File: examples/parallelize_mlp.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  def data_stream():
+    for i in range(num_batches):
+      batch_idx = slice(i * batch_size, (i + 1) * batch_size)
+      yield train_images[batch_dx], train_labels[batch_idx]
+  batches = data_stream()
+
+  _, params = init_random_params((-1, 28 * 28))
+  loss_ = partial(loss, params)
+  spmd_loss, axis_name = papply(loss_, train_images.shape[0])
+  pmap_loss = serial_pmap(spmd_loss, axis_name)
```
Comment (mattjj) [inline examples/parallelize_mlp.py]: Since this is just using `serial_pmap` at the moment, maybe leave a TODO comment here, and/or change the description at the top to explain that it's not actually parallelized yet?

In fact, it might be worth considering not merging this example into master yet until we make it work with `pmap`, since WIP code probably shouldn't be in the canonical "examples."

I'm split on that. But I expect if the example doesn't go in now, it'll go in over the next several days! Up to you.

**Review by mattjj — APPROVED** [review]
Awesome!

I put a question about whether to merge the example now. Otherwise LGTM.

File: examples/parallelize_mlp.py (hunk 1 above)
Comment (froystig) [reply examples/parallelize_mlp.py]: Yes, good point. I've removed the example for now.
